src/coredump/coredump.c

```diff
@@ -810,49 +827,51 @@ static int submit_coredump(
         return 0;
 }
 
-static void map_context_fields(const struct iovec *iovec, const char* context[]) {
-
-        static const char * const context_field_names[] = {
-                [CONTEXT_PID] = "COREDUMP_PID=",
-                [CONTEXT_UID] = "COREDUMP_UID=",
-                [CONTEXT_GID] = "COREDUMP_GID=",
-                [CONTEXT_SIGNAL] = "COREDUMP_SIGNAL=",
-                [CONTEXT_TIMESTAMP] = "COREDUMP_TIMESTAMP=",
-                [CONTEXT_RLIMIT] = "COREDUMP_RLIMIT=",
-                [CONTEXT_HOSTNAME] = "COREDUMP_HOSTNAME=",
-                [CONTEXT_COMM] = "COREDUMP_COMM=",
-                [CONTEXT_EXE] = "COREDUMP_EXE=",
-        };
+static int cache_process_info(ProcessInfoCache *cache, const struct iovec_wrapper *iovw) {
+        unsigned n, i, count = 0;
+        char *unit;
+        int r;
 
-        unsigned i;
+        assert(cache);
+        assert(iovw);
+        assert(iovw->count >= _META_ARGV_MAX);
 
-        assert(iovec);
-        assert(context);
+        /* The cache does not allocate any memory on its own */
 
-        for (i = 0; i < ELEMENTSOF(context_field_names); i++) {
-                char *p;
+        memzero(cache, sizeof(ProcessInfoCache));
 
-                if (!context_field_names[i])
-                        continue;
+        for (n = 0; n < iovw->count; n++) {
+                struct iovec *iovec = iovw->iovec + n;
 
-                p = memory_startswith(iovec->iov_base, iovec->iov_len, context_field_names[i]);
-                if (!p)
-                        continue;
+                for (i = 0; i < ELEMENTSOF(meta_field_names); i++) {
+                        char *p;
 
-                /* Note that these strings are NUL terminated, because we made sure that a trailing NUL byte is in the
-                 * buffer, though not included in the iov_len count. (see below) */
-                context[i] = p;
-                break;
+                        p = startswith(iovec->iov_base, meta_field_names[i]);
+                        if (p) {
+                                cache->meta[i] = p;
+                                count++;
+                                break;
+                        }
+                }
         }
+
+        r = parse_pid(cache->meta[META_ARGV_PID], &cache->pid);
```

Comment:
> This needs a check that cache->meta[META_ARGV_PID] != NULL

Indeed, if the function has been called by process_socket() then we can be sure that all metadata have been sent through the socket. OTOH if the metadata have been read by us, we're sure that all mandatory metadata are present.

Reply:
> I think it'd also be nicer to move the check for mandatory fields to this function too.

No because they're read in 2 steps (gather_pid_metadata_from_argv and gather_pid_metadata_from_runtime) in case the metadata are retrieved by the coredump handler.

Reply:
> Also, `cache->meta[META_RT_UNIT]` gets its own helper variable, but `cache->meta[META_ARGV_PID]`. This asymmetry is not nice.

META_RT_UNIT isn't cached because it's not needed by the rest of the code where the pid of the process is.